guide.md

```diff
@@ -0,0 +1,350 @@
+rules_nixpkgs_dependencies()
+```
+
+Now we can select a version of the Nixpkgs repository from which all of our
+packages will be drawn. Any Git commit hash or tag is allowed. This way, we
+select a consistent set of package revisions in one go instead of having to
+maintain dependency versions for individual dependencies. Often this will be a
+tag for a stable release, which ensures that there has been especially
+stringent verification that packages in it will function correctly individually
+and play well together. Here we use the 22.05 release:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_git_repository")
+nixpkgs_git_repository(
+    name = "nixpkgs",
+    revision = "22.05",
+    sha256 = "",
+)
+```
+
+# A hermetic CC toolchain
+
+We can now define a CC toolchain that will be provisioned by Nix using
+package definitions from the 22.05 release of Nixpkgs.
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  name = "nixpkgs_config_cc",
+)
+```
+
+This will select the default `stdenv.cc` package as the compiler, drawing
+auxiliary tools from `stdenv.cc.bintools`. However, one can write a little bit of
+Nix to configure this:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  # nix_file_content = "(import <nixpkgs> {}).gcc11",
+  attribute_path = "gcc11",
+  name = "nixpkgs_config_cc",
+)
+```
+
+`(import <nixpkgs> {})` loads the entire package set into an attribute set
+(Nix's equivalent of a record), and `.gcc11` selects one value from it. It is
+possible to customize the compiler further using more Nix if desired: you could,
+for instance, point Nix to a custom fork of Clang containing support for a new
+architecture, and have everything work seamlessly.
+
+With this, we now have a fully hermetic CC toolchain that will behave correctly
+across platforms. It remains to tell Bazel to use it by default, instead of
+`@local_config_cc`. To do so, we put the following into `.bazelrc` at the root
+of the repository:
+
+```
+build --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+Now `bazel build` and `bazel run` should use this toolchain by default for
+C/C++ compilation:
+```
+> bazel build //src:hello-world
+INFO: Analyzed target //src:hello-world (42 packages loaded, 177 targets configured).
+INFO: Found 1 target...
+Target //src:hello-world up-to-date:
+  bazel-bin/src/hello-world
+INFO: Elapsed time: 0.377s, Critical Path: 0.22s
+INFO: 6 processes: 4 internal, 2 linux-sandbox.
+INFO: Build completed successfully, 6 total actions
+```
+
+# Making Nix optional
+
+It is entirely possible to make the use of Nix optional in a Bazel setup. To do
+so, we use `bazelrc` configs to only enable Nix-specific configuration when a
+`--config` flag is provided, like so:
```

Comment:
Hadn't thought about that, I'll make a note of this. 

As for defining a non-Nix toolchain, the `rules_cc_toolchains()` call takes care of that, right?

Reply:
I've added a note about this; could you check that it makes sense? (I wasn't really sure what the "default bindist" specifically referred to.)

Reply:
Thanks, yes, that makes sense. The exact shape of the "default bindist" depends on the language. With cc its implicit since it's builtin to Bazel. With other languages it would be your `rules_XXX_toolchains` call, or similar that sets this up, e.g. [here for Go](https://github.com/tweag/rules_nixpkgs/blob/56b9cff9d175b916abdb36920a669b1face49e04/examples/toolchains/go/WORKSPACE#L45).